### PR TITLE
ARROW-3793: [C++] TestScalarAppendUnsafe is not testing unsafe appends

### PR DESCRIPTION
**Pull request by bkietz**
Perform enough repetitions that the Buffers will be more than the minimum allocation size (64 bytes) and reserve enough space for all repetitions. Also use UnsafeAppendNull rather than AppendNull to ensure that nothing is calling Reserve in the test's loop

### CODE REVIEW

**Comment by wesm** [comment]
We have a bunch of warnings enabled in Travis CI, can be done with `-DBUILD_WARNING_LEVEL=CHECKIN`

```
/home/travis/build/apache/arrow/cpp/src/arrow/array-test.cc:1504:41: error: conversion to ‘int’ from ‘std::basic_string<char>::size_type {aka long unsigned int}’ may alter its value [-Werror=conversion]
   for (auto&& s : strings) total_length += s.size();
```

**Comment by wesm** [comment]
Ah, the breaks should be put in the else part of the ifdef

**Comment by bkietz** [comment]
I could add that to this PR

**Comment by bkietz** [comment]
I'm on a fresh Ubuntu 18.04 VM, compiling with clang-6.0. I used conda to get my dependencies:

<details>
<pre>
# packages in environment at /home/bkietz/miniconda3/envs/arrow-dev:
#
# Name                    Version                   Build  Channel
boost-cpp                 1.68.0               h3a22d5f_0    conda-forge
brotli                    1.0.7                hfc679d8_0    conda-forge
bzip2                     1.0.6                h470a237_2    conda-forge
ca-certificates           2018.10.15           ha4d7672_0    conda-forge
certifi                   2018.10.15            py37_1000    conda-forge
cmake                     3.12.4               h011004d_0    conda-forge
curl                      7.62.0               h74213dd_0    conda-forge
expat                     2.2.5                hfc679d8_2    conda-forge
flatbuffers               1.10.0               hfc679d8_0    conda-forge
gettext                   0.19.8.1             h5e8e0c9_1    conda-forge
gflags                    2.2.2                hfc679d8_0    conda-forge
git                       2.19.1          pl526hbb17d3c_1    conda-forge
icu                       58.2                 hfc679d8_0    conda-forge
krb5                      1.16.2               hbb41f41_0    conda-forge
libcurl                   7.62.0               hbdb9355_0    conda-forge
libedit                   3.1.20170329         haf1bffa_1    conda-forge
libffi                    3.2.1                hfc679d8_5    conda-forge
libgcc-ng                 8.2.0                hdf63c60_1  
libiconv                  1.15                 h470a237_3    conda-forge
libssh2                   1.8.0                h5b517e9_2    conda-forge
libstdcxx-ng              7.2.0                hdf63c60_3    conda-forge
libuv                     1.24.0               h470a237_0    conda-forge
lz4                       2.1.2            py37h7b6447c_0  
lz4-c                     1.8.3                hfc679d8_0    conda-forge
make                      4.2.1             h470a237_1002    conda-forge
ncurses                   6.1                  hfc679d8_1    conda-forge
ninja                     1.8.2                h2d50403_1    conda-forge
openssl                   1.0.2p               h470a237_1    conda-forge
perl                      5.26.2               h470a237_0    conda-forge
pip                       18.1                  py37_1000    conda-forge
python                    3.7.1                h5001a0f_0    conda-forge
rapidjson                 1.1.0                         0    conda-forge
readline                  7.0                  haf1bffa_1    conda-forge
rhash                     1.3.6                h470a237_1    conda-forge
setuptools                40.6.2                   py37_0    conda-forge
snappy                    1.1.7                hfc679d8_2    conda-forge
sqlite                    3.25.3               hb1c47c0_0    conda-forge
thrift-cpp                0.11.0               he11dc3d_3    conda-forge
tk                        8.6.8                ha92aebf_0    conda-forge
wheel                     0.32.2                   py37_0    conda-forge
xz                        5.2.4                h470a237_1    conda-forge
zlib                      1.2.11               h470a237_3    conda-forge
zstd                      1.3.7                h502d103_0    conda-forge
</pre>
</details>

I hit this because I didn't set `ARROW_BUILD_BZ2` when building.

**Comment by bkietz** [comment]
The test is failing because `UnsafeAppendToBitmap()` only modifies length and null count, but `BinaryBuilder` needs that to append 0 to its offsets buffer. Could fix this for example by making `UnsafeAppendToBitmap()` virtual and overriding it in `BinaryBuilder`, what do you think @wesm ?

**Comment by wesm** [comment]
How about adding an UnsafeAppendNull method? These methods are performance sensitive and so it probably wouldn't make sense to make virtual

**Comment by bkietz** [comment]
If we add UnsafeAppendNull to BinaryBuilder then that won't fix this bug when the dynamic type is BinaryBuilder but we are referencing it through a pointer to ArrayBuilder:

```
BinaryBuilder *bb = get();
ArrayBuilder *ab = bb;
ab->UnsafeAppendNull(); // doesn't modify offsets
```

I assumed this is a case we need to worry about because ArrayBuilder has a virtual destructor. If not then I suggest we add a warning to the doccomment that builders need to be referenced through maximally downcast pointers.

**Comment by wesm** [comment]
Using a pointer to base isn't really a supported API at the moment so it's fine to presume that people will be using a pointer to the actual builder type

**Comment by bkietz** [comment]
Is that pattern documented? I feel like users could easily be surprised by this and we should be as explicit as possible.

Alternatively, if we made `ArrayBuilder::UnsafeAppendNull` protected, then subclasses could either `using` them up to public or implement their own versions.

**Comment by wesm** [comment]
If you make this change you'll have to add `UnsafeAppendNull()` at least for `NumericBuilder<T>` also because it's used in `arrow::csv` https://github.com/apache/arrow/blob/df54da211448b5202aa08ed2b245eb78cfd1e50c/cpp/src/arrow/csv/converter.cc#L297

File: cpp/src/arrow/builder.h [hunk 1]
```diff
@@ -370,6 +372,11 @@ class ARROW_EXPORT NumericBuilder : public PrimitiveBuilder<T> {
           ARROW_MEMORY_POOL_DEFAULT)
       : PrimitiveBuilder<T1>(TypeTraits<T1>::type_singleton(), pool) {}
 
+  using ArrayBuilder::AppendToBitmap;
+  using ArrayBuilder::SetNotNull;
+  using ArrayBuilder::UnsafeAppendNull;
+  using ArrayBuilder::UnsafeAppendToBitmap;
```
Comment (wesm) [inline cpp/src/arrow/builder.h]: Can you make only UnsafeAppendNull public? It looks like there's one place where `UnsafeAppendToBitmap(false)` will have to be replaced with `UnsafeAppendNull`

File: cpp/src/arrow/builder.h [hunk 2]
```diff
@@ -859,6 +867,8 @@ class ARROW_EXPORT BinaryBuilder : public ArrayBuilder {
 
   Status AppendNull();
 
+  void UnsafeAppendNull();
```
Comment (wesm) [inline cpp/src/arrow/builder.h]: Maybe put the implementation in the header (so it's inlined)?

**Comment by wesm** [comment]
The appveyor failure looks spurious to me

**Review by wesm — APPROVED** [review]
+1, thanks @bkietz!

**Pull request closed without merging**